### PR TITLE
feat: Introduce `NoEmptyBlockFixer`

### LINKED ISSUE #3514: Empty block fixer
I would like to have a fixer that would remove statements with empty blocks.

Example:

```diff
-if ($foo) {
-}

-while ($bar) {
-}

-switch ($baz) {
-}
```

### PR DESCRIPTION
**Pull request by Wirone**
Fixes #3514, [created](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/3514#issuecomment-393839308) by @ntzm, rebased and synced with current requirements by me 😉.

1 test case is failing, I will investigate only after initial green light that we actually want this fixer (@keradus [said](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/3514#issuecomment-393854206) it's OK if it's marked as risky, and it is).

### CODE REVIEW

**Comment by Wirone** [comment]
FYI: PHPStorm reports duplicated code fragment:

image

**Comment by ntzm** [comment]
Wow, I never thought anyone would use my branch for anything, nice one :D

File: src/Fixer/ControlStructure/NoEmptyBlockFixer.php [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+namespace PhpCsFixer\Fixer\ControlStructure;
+
+use PhpCsFixer\AbstractFixer;
+use PhpCsFixer\FixerDefinition\CodeSample;
+use PhpCsFixer\FixerDefinition\FixerDefinition;
+use PhpCsFixer\FixerDefinition\FixerDefinitionInterface;
+use PhpCsFixer\Tokenizer\Tokens;
+
+/**
+ * @author Nat Zimmermann <nat@razorcreations.com>
```
Comment (ntzm) [inline src/Fixer/ControlStructure/NoEmptyBlockFixer.php:24]: Can you change this to `nathanielzimmermann@gmail.com` please? This is an email address for a previous employer that I accidentally used. Or perhaps remove the email entirely if it's not needed.

Comment (Wirone) [reply src/Fixer/ControlStructure/NoEmptyBlockFixer.php:24]: I took it from the commits, so I need to amend author too, will do it later.

Comment (ntzm) [reply src/Fixer/ControlStructure/NoEmptyBlockFixer.php]: Thank you!

File: src/Fixer/ControlStructure/NoEmptyBlockFixer.php [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+
+/**
+ * @author Nat Zimmermann <nathanielzimmermann@gmail.com>
+ */
+final class NoEmptyBlockFixer extends AbstractFixer
+{
+    public function getDefinition(): FixerDefinitionInterface
+    {
+        return new FixerDefinition(
+            'There must not be any empty blocks.',
```
Comment (kubawerlos) [inline src/Fixer/ControlStructure/NoEmptyBlockFixer.php:31]:
```suggestion
            'There must be no empty blocks.',
```

File: doc/rules/control_structure/no_empty_block.rst [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+   \ No newline at end of file
+
+Example #3
+~~~~~~~~~~
+
+.. code-block:: diff
+
+   --- Original
+   +++ New
+   -<?php while ($foo) {}
```
Comment (keradus) [inline doc/rules/control_structure/no_empty_block.rst:47]: `while($itrator->consume()) {}`

I wonder if it makes sense to have config which structures could be removed

File: tests/Fixer/ControlStructure/NoEmptyBlockFixerTest.php [hunk 4]
```diff
@@ -0,0 +1,815 @@
...
+            '<?php ',
+            '<?php if ($foo): elseif ($bar): elseif ($baz): elseif ($boz): endif;',
+        ];
+
+        yield 'alternate multiple elseif one with side effect' => [
+            '<?php if ($foo): elseif ($bar): foo(); elseif ($baz): elseif ($boz): endif;',
+        ];
+
+        yield 'alternate if with side effect in else' => [
+            '<?php if ($foo): else: bar(); endif;',
```
Comment (keradus) [inline tests/Fixer/ControlStructure/NoEmptyBlockFixerTest.php:240]: fyi
i remember we had the discussion that we may stop touching any file that is using alternative syntax one day

File: tests/Fixer/ControlStructure/NoEmptyBlockFixerTest.php [hunk 5]
```diff
@@ -0,0 +1,815 @@
...
+        ];
+
+        yield 'if with else without side effect with comments' => [
+            '<?php /*1*//*2*//*3*//*4*//*5*//*6*//*7*//*8*/',
+            '<?php /*1*/if/*2*/(/*3*/$foo/*4*/)/*5*/{}/*6*/else/*7*/{}/*8*/',
+        ];
+
+        yield 'if with else without side effect with doc comments' => [
+            '<?php /**1*//**2*//**3*//**4*//**5*//**6*//**7*//**8*/',
+            '<?php /**1*/if/**2*/(/**3*/$foo/**4*/)/**5*/{}/**6*/else/**7*/{}/**8*/',
```
Comment (keradus) [inline tests/Fixer/ControlStructure/NoEmptyBlockFixerTest.php:138]:
```suggestion
            '<?php /**1*/if/**2*/(/**3*/$foo/**4*/)/**5*/{}/**6*/else/**7*/{/*inside*/}/**8*/',
```

File: tests/Fixer/ControlStructure/NoEmptyBlockFixerTest.php [hunk 6]
```diff
@@ -0,0 +1,815 @@
...
+     * @param null|string $input
+     *
+     * @dataProvider provideTestFixCases
+     */
+    public function testFix($expected, $input = null): void
+    {
+        $this->doTest($expected, $input);
+    }
+
+    public static function provideTestFixCases(): iterable
```
Comment (keradus) [inline tests/Fixer/ControlStructure/NoEmptyBlockFixerTest.php:39]:
```php

<?php

echo 1;

{}

echo 2;
```

File: tests/Fixer/ControlStructure/NoEmptyBlockFixerTest.php (hunk 4 above)
Comment (Wirone) [reply tests/Fixer/ControlStructure/NoEmptyBlockFixerTest.php:240]: Yes, but there are many other fixers with test cases for alternative syntax so I think we can leave it here too. If we finally decide to drop such support, we can then remove it from all fixers.

File: tests/Fixer/ControlStructure/NoEmptyBlockFixerTest.php (hunk 6 above)
Comment (Wirone) [reply tests/Fixer/ControlStructure/NoEmptyBlockFixerTest.php:39]: `Fixer must be a candidate for input code.` - currently fixer works for loops, switches etc. Is it required to handle this case?

File: doc/rules/control_structure/no_empty_block.rst (hunk 3 above)
Comment (Wirone) [reply doc/rules/control_structure/no_empty_block.rst:48]: It can be added later as improvement, I would rather leave it as is now.

File: tests/Fixer/ControlStructure/NoEmptyBlockFixerTest.php (hunk 4 above)
Comment (keradus) [reply tests/Fixer/ControlStructure/NoEmptyBlockFixerTest.php:252]: it's more like "consider to not bother about it next time"

File: tests/Fixer/ControlStructure/NoEmptyBlockFixerTest.php (hunk 6 above)
Comment (keradus) [reply tests/Fixer/ControlStructure/NoEmptyBlockFixerTest.php]: I would imagine `{}` to be handled by `NoEmptyBlockFixer`, if we want to touch only `loop`, then we would have `NoEmptyBodyBlockOfLoopFixer, but I would recommend to not have that very specialized rule that covers only some limited cases

Comment (Wirone) [reply tests/Fixer/ControlStructure/NoEmptyBlockFixerTest.php]: I agree, I just don't think that rule must cover every possible scenario since the very beginning. It's not my code, I just refurbished it from the ancient branch, I would like to provide it "as-is" (not counting required alignments). Then we can improve it even more. Example with `{}` is such an edge case that in my opinion doesn't have to be handled now.

**Comment by Wirone** [comment]
@ntzm do you have an idea why nested `if` test case fails? I did not have time to dig it yet, maybe you can help make pipeline green?

**Comment by ntzm** [comment]
Not sure sorry, I don't really have any capacity nowadays to work on anything PHP open source related 😢

**Comment by SpacePossum** [comment]
As it fixes `if` and `else` blocks, should this fix `elseif` blocks as well?

**Comment by Wirone** [comment]
@SpacePossum please look [here](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7061#discussion_r1254982555) 😉. I've rebased it only without any additional changes, will try to solve CI issues soon but I don't want to work on this PR more than it's required. Any improvements can be added later.

**Comment by SpacePossum** [comment]
ok, maybe tag is as draft than?

**Comment by Wirone** [comment]
Personally I consider something as draft when it is not finished in terms of conception, so potential reviewers would know there might be any changes provided. This is ~final state of the PR, it just does not pass CI and requires fixes 😉.

**Comment by SpacePossum** [comment]
That confuses me even more : D   
if the PR is kinda final, than I stick with my questions, I know you revived an older work but that doesn't mean questions cannot be done?

**Comment by Wirone** [comment]
No, it means that there's no much point in reviewing draft, unless author explicitly asks for the feedback (personally I ignore draft MRs, more [here](https://blog.codito.dev/2022/07/code-review-good-practices/#draft-wip)). That's why I use draft state only when it's really under active development and everything may change in any direction. Failed CI is next-step indicator that MR requires work, but it suggests at least that the code is ready for review and won't change much 🙂. Personally I briefly look at such MRs and give initial feedback, waiting for green pipeline.

**Comment by SpacePossum** [comment]
ok so we wait for green CI and than you'll answer the reviewer questions before merging?

**Comment by Wirone** [comment]
@SpacePossum which question of yours I did not respond to 🤷‍♂️? In terms of empty `{}` I just redirected you to other thread where I already said I don't want to support it. It's not my code, I just refreshed @ntzm's work and I try to make it pass CI in my spare time. Then I will make final code review myself, as I feel more like reviewer here. I don't get what do you want from me, really 🙂.

**Comment by SpacePossum** [comment]
Not sure about a "problem," but this PR doesn't look finished to me, no matter who was or is the author, so I can ask no?
I tried something here
https://github.com/Wirone/PHP-CS-Fixer/pull/3
take or leave, really strange talk, feel free to tag me "off topic" again :)

**Comment by Wirone** [comment]
It is not finished in terms of CI status. It may look unfinished in terms of feature coverage and feel free to leave any comment related to current code. I'll answer and/or apply changes as soon as I can. In a meantime I try to just keep my PRs on top of the master, it does not mean I actively work on this 😉. And yeah, this whole discussion was off-topic.

**Comment by Wirone** [comment]
I really need to get back to this and try to finish it 😅. Sorry folks for the long delay!

File: tests/Fixer/ControlStructure/NoEmptyBlockFixerTest.php
Comment (julienfalque) [inline tests/Fixer/ControlStructure/NoEmptyBlockFixerTest.php:1]: According to test cases, it looks like control structure conditions are only considered to have side effects when calling a function. I wonder if we should consider more scenarios, e.g.:
* assignations, e.g. `$foo = $bar`;
* reading an index, e.g. `$foo[$index]`: `$foo` might be an `ArrayAccess` implementation, so the code might have side effects;
* some operations, e.g. `$foo + $bar`: operator overloading isn't possible directly in PHP code but it is possible in PHP extensions, which means such operations might have side effects.
